src/hammer-vlsi/hammer_vlsi/constraints.py

```diff
@@ -242,42 +244,138 @@ class PlacementConstraint(NamedTuple('PlacementConstraint', [
     __slots__ = ()
 
     @staticmethod
-    def from_dict(constraint: dict) -> "PlacementConstraint":
+    def from_dict(masters: List[MacroSize], constraint: dict) -> "PlacementConstraint":
         constraint_type = PlacementConstraintType.from_str(
             str(constraint["type"]))
+
+        ### Margins ###
+        # This field is mandatory in TopLevel constraints
+        # This field is disallowed otherwise
```

Comment:
Can we move this to defaults.yml for DRY?

Reply:
I will also put it in defaults.yml, but it was very helpful in writing the code (and reading it after writing), so I would rather not delete it.